vesc_hw_interface/src/vesc_step_difference.cpp

```diff
@@ -0,0 +1,134 @@
+ * the previous step as is.
+ */
+void VescStepDifference::enableSmooth(double control_rate, double max_sampling_time, int max_step_diff)
+{
+  enable_smooth_ = true;
+  step_diff_vw_max_step_ = max_step_diff;
+  step_diff_vw_max_window_size_ = std::max(1, static_cast<int>(control_rate * max_sampling_time));
+  step_input_queue_.resize(step_diff_vw_max_window_size_ + 1);
+  return;
+}
+
+/**
+ * Outputs the difference between the previous input and the current input.
+ * Smoothing is performed when "enable_smooth_" is true
+ */
+double VescStepDifference::getStepDifference(const double step_in, bool reset)
+{
+  double output;
+  if (enable_smooth_)
+  {
+    output = this->stepDifferenceVariableWindow(step_in, reset);
+  }
+  else
+  {
+    output = this->stepDifferenceRaw(step_in, reset);
+  }
+  return output;
+}
+
+/**
+ * Function called when the filter is disabled
+ *
+ * Outputs the difference of the counter value from the previous value.
+ * This function is used when no filter is used.
+ *
+ * @step_in Enter the current counter value
+ * @reset Used for initialization.
+ * @return Returns the difference from the previous value.
+ */
+double VescStepDifference::stepDifferenceRaw(const double step_in, bool reset)
+{
+  if (reset)
+  {
+    step_diff_raw_prev_ = static_cast<int16_t>(step_in);
+    return 0.0;
+  }
+  int16_t step_diff = static_cast<int16_t>(step_in) - step_diff_raw_prev_;
+  double output = static_cast<double>(step_diff);
+  step_diff_raw_prev_ = static_cast<int16_t>(step_in);
+  return output;
+}
+
+/**
+ * Function called when the filter is enabled
+ *
+ * Output smoothed differences.
+ * The smaller the difference, the longer the sampling time is stretched and smoothed.
+ * If the latest difference is greater than max_step_diff, no smoothing is performed.
+ * Smoothing is performed for max_sampling_time seconds when the latest difference is 0.
+ *
+ * @step_in Enter the current counter value
+ * @reset Used for initialization.
+ * @return Returns the difference from the previous value.
+ */
+double VescStepDifference::stepDifferenceVariableWindow(const double step_in, bool reset)
+{
+  if (reset)
+  {
+    // Fill the buffer with the current value
+    std::fill(step_input_queue_.begin(), step_input_queue_.end(), static_cast<int16_t>(step_in));
+    return 0.0;
+  }
+  // Increment buffer
+  step_input_queue_.pop_back();
+  step_input_queue_.push_front(static_cast<int16_t>(step_in));
+  // Calculate window size
+  int latest_step_diff = std::abs(step_input_queue_[0] - step_input_queue_[1]);
+  int window_size = step_diff_vw_max_window_size_;
+  if (latest_step_diff >= step_diff_vw_max_step_)
+  {
+    window_size = 1;
+  }
+  else if (latest_step_diff > 0)
+  {
+    window_size = step_diff_vw_max_window_size_ *
+                  (1.0 - static_cast<double>(latest_step_diff) / static_cast<double>(step_diff_vw_max_step_));
```

Comment:
`window_size` implicitly apply `floor()` to convert to `int`. Is it correct? Should be round()? If it is correct, please explicitly apply `std::floor`.

Reply:
The decimal point handling is corrected to use std::round().
There was an expression with a misaligned queue element.
I'll fix it and commit.